### PR TITLE
#563 cleanup ReduceStateHolder<T> at VT finalize

### PR DESCRIPTION
**Pull request by lifflander**

### CODE REVIEW

File: src/vt/collective/reduce/reduce_state_holder.impl.h [hunk 1]
```diff
@@ -95,20 +95,31 @@ template <typename T>
...
+/*static*/ void ReduceStateHolder<T>::cleanup() {
+  state_lookup_.clear();
 }
 
 template <typename T>
 /*static*/ typename ReduceStateHolder<T>::GroupLookupType
 ReduceStateHolder<T>::state_lookup_ = {};
 
+template <typename T>
+/*static*/ bool ReduceStateHolder<T>::created_ = false;
```
Comment (PhilMiller) [inline src/vt/collective/reduce/reduce_state_holder.impl.h]: I would reverse the sense of this, so it doesn't need to be negated later. Something like `first_insertion_ = true`

File: src/vt/collective/reduce/reduce.impl.h [hunk 2]
```diff
@@ -163,7 +163,15 @@ void Reduce::reduceAddMsg(
     ReduceState<MessageT> state(
       msg->reduce_tag_,msg->reduce_seq_,num_contrib_state
     );
-    ReduceStateHolder<MessageT>::insert(group_,lookup,std::move(state));
+    bool has_elm = ReduceStateHolder<MessageT>::insert(group_,lookup,std::move(state));
+    // For the first time of type MessageT, add a cleanup function to clear out
+    // messages. They must be cleaned up early so message destructors do not run
+    // after VT is shutdown.
+    if (not has_elm) {
+      cleanup_.emplace_back([]{
```
Comment (pnstickne) [inline src/vt/collective/reduce/reduce.impl.h]: Functions in templates come at cumulatively expensive T-cost.

**Comment by lifflander** [comment]
I think this problem of messages being held in the `ReduceStateHolder<T>` past finalize is actually a symptom of another bug. Thus, I am holding off on merging this.

**Comment by lifflander** [comment]
Closing this as the problem was a consequence of other bugs and is not a real problem

**Pull request closed without merging**